### DISCUSSION TITLE
Two-Layer Design Model Proposal

### DISCUSSION THREAD

**Discussion by talltree** [trustoverip/trust-spanning-protocol#43] [General]
# Overview

This is a proposal for consideration at the consolidation stage of the Trust Spanning Protocol TF work. The purpose of this proposal is to see if we might resolve some of the tensions between the initial design proposals by separating the ToIP Layer 2 trust spanning protocol (TSP) from a protocol at the layer immediately above it **but at the bottom of ToIP Layer 3, i.e., below what we have typically considered distinct ToIP trust tasks**. For purposes of this proposal only we will call this the **trust task messaging protocol (TTMP)** (if we end out going this direction, we can discuss later what is the best name for this protocol).

# Why Separate the TTMP from the TSP?

In the initial proposal stage, two of the proposals (# 1 from Sam Smith and # 3 from Wenjing Chu) emphasized the need for the TSP to be as lightweight and minimal as possible. Besides the general requirement for any spanning protocol to be "as simple as possible but no simpler", the two main reasons given for this were:

1. **Security.** The simpler the protocol, the fewer attack surfaces, and the stronger the security analysis (and ideally a formal method proof of security).
2. **Connectivity/interoperability.** The simpler the protocol, the wider range of systems that will be able to use it to interconnect across trust domains.

By contrast, proposals # 2 from Daniel Hardman and # 4 from Michael Herman emphasized the need for the TSP to provide the necessary inter-trust-application messaging support for a wide range of trust tasks at ToIP Layer 3. Many of these trust tasks will benefit from having common semantics established at a lower layer so they do not have to re-invent the same messaging properties (in potentially non-interoperable ways) to accomplish their work. As an example, many of these trust task messaging semantics are incorporated into [the DIDComm V2 specification](https://identity.foundation/didcomm-messaging/spec/v2.0/).

# Is this similar to the separation in the TCP/IP stack of the TCP and UDP layer from the IP layer?

As an _analogy_, yes. However, this is primarily an analogy about the relative _scopes_ of the protocols at those two layers, not their specific functionality. Here is a quote from [section 1.2 of the original Internet Protocol RFC 791](https://www.rfc-editor.org/rfc/rfc791#section-1.2) (September 1981):

> The internet protocol is specifically limited in scope to provide the functions necessary to deliver a package of bits (an internet datagram) from a source to a destination over an interconnected system of networks. There are no mechanisms to augment end-to-end data reliability, flow control, sequencing, or other services commonly found in host-to-host protocols. The internet protocol can capitalize on the services of its supporting networks to provide various types and qualities of service.

This illustrates how the architects applied the spanning layer “simple as possible but no simpler” design principle. Note the phrase “host-to-host protocols”. That’s how the TCP and UDP layer is described, whereas the IP protocol is strictly about addressing and package (datagram) delivery.

# What would be the primary benefit to the TSP of separating out the TTMP?

By separating out the higher-level semantics of trust task messaging, the design of the TSP could be as minimal as possible (as advocated by Wenjing) and guided by the PAC trilemma and the Layered Security Overlay Spiral (as advocated by Sam Smith in [this paper](https://github.com/SmithSamuelM/Papers/blob/master/whitepapers/SPAC_Message.md) and discussed in [this Github discussion thread](https://github.com/trustoverip/trust-spanning-protocol/discussions/33)).

# What would be the primary benefit to the TTMP of layering over the TSP?

If the TSP handles authenticity, confidentiality, privacy (and availability?), then the design of the TTMP can be simplified because it only needs to focus on adding the shared messaging semantics required by many higher level trust tasks. In this case, the TTMP should in theory be able to leverage the 5+ years of work on DIDComm V2 (as advocated by Michael Herman and Sam Curren) — including the composability of higher layer trust task protocols (as advocated by Michael Herman and Daniel Hardman).

This layering of a higher-level messaging protocol over the TSP might also have the benefit of enabling other protocols at that layer that implement specialized messaging patterns (group messaging? broadcast messaging?). Those protocols could benefit from running over the shared TSP layer the same way TCP and UDP implement different host-to-host messaging patterns running over the shared IP layer.

# What if this approach completely fails?

It is entirely possible that this line of thinking is hopelessly flawed, and that the TSP should handle all common requirements of the trust task protocol layer above it. If that proves to be true, what discussion of this proposal could teach us is **why** a single layer model makes more sense than a two-layer model. That in turn should help us with the design of the TSP.

**Comment by dhh1128** [comment]
Interesting proposal. Well worth considering.

Since we began this task force, here are some ways that a needle has moved in my brain:

* I have shifted from being focused on headers to accepting Sam's idea of overlays.
* I have seen how Sam's ESSR model may be more desirable than DIDComm's wrapping strategy. (Pending some details to be sorted.)
* I have accepted the desirability of [appraisable](https://github.com/trustoverip/trust-spanning-protocol/discussions/41#discussioncomment-5586892) {authenticity, confidentiality, privacy} as a way to resolve the tension between the AID requirement in Sam's proposal and mine, and Wenjing's preference to be less opinionated about identifiers.

All of these feel like important moves toward alignment. Maybe this proposal will trigger additional moves in the right direction. :-)

However, to me, characterizing a proposal in terms of where it draws lines might be emphasizing the wrong issue. There are lots of ways to draw lines, and I could live with many of them. What's more interesting is: *what do those lines imply about dependencies, cohesion, and requirements?* If TTMP is separated by a line from TSP, when would we strongly intend to advocate that people should use TTMP? And when would we shrug our shoulders and say we have no skin in the game, no opinion on the subject, and are agnostic with respect to whether the TTMP's trust goals are important and its methods useful? *Those* are the questions. I've argued about lines because I see the lines as indicators of our attitude about those questions: it feels like the intention is to deliver "core trust" with what's on this side of the line, and auxiliary and higher-level concerns on the other side. And I don't like our definition of "core trust."

Wenjing has repeatedly said that by ruling something out of scope, he's not claiming it's unimportant. I hear that. But I believe there is a strong coupling between the concerns he says are in scope for TSP, and some concerns that he says are out of scope. If I am right, then drawing a line to divide the concerns into two groups might still be useful for political or procedural reasons, but it is invalid *from a dependencies, cohesion, and requirements* perspective. It sends the wrong signal to the market, and teaches the wrong mental model about what is actually a sufficient foundation for trust.

So: to the extent that we imagine advocating strongly for the functionality that Drummond fits into TTMP, I tend to feel more cheerful about this proposal; to the extent that we intend to be agnostic, I tend to feel less cheerful.

You might think I feel that way because I want DIDComm to have more champions on this TF. If so, you're in the right ballpark, but maybe not exactly accurate. :-) What I actually care about is that we define "trust" inclusively enough, and that we provide a very strong foundation for it. I want whatever is considered foundational -- however we draw the lines and however we encapsulate the concerns -- to add to its scope of concerns some additional primitives that fall outside the realm of cryptographic mechanisms and PAC tradeoffs.

Defining how to pass messages from *identifier* A to B (as opposed to *IP address* A to B), while providing cryptographic guarantees of authentic origin, is an excellent beginning of what we need. It is a substantial improvement on the status quo. If we provide a certain set of confidentiality guarantees on top of this, and if we allow for optional privacy (e.g., whatever subset of Sam's ESSR thinking that we decide fits in the base), we are getting closer, but I feel that *we are still not there*. Instead of being "as narrow as possible, but no narrower", we are still "too narrow." The piece that I perceive to be missing is what I've been advocating that we also provide.

What I want is a future where ToIP is loud and proud that A) those missing things are also essential to any sophisticated trust strategy; B) there's a canonical way to achieve them. If we accomplish A and B, and those things are in a different layer from a some other layer, then I can be happy. But to the extent that we create the extra layer only to make the extra layer an optional thing, and a peer to current or future constructs from arbitrary inventors that want to think about trust differently, IMO we are not building something that deserves the name "trust spanning protocol." We are instead providing a "few key ingredients of trust" protocol.

These missing things are what I called "authentic intentions" in my proposal. I described them in the abstract on [slide 26](https://docs.google.com/presentation/d/1V3v3-g6pAR8oMhk0Az-pheFM7PRCxJO9OsonH7STi5I/edit#slide=id.g20885a66385_0_26), gave a concrete example of how we might solve for them on [slide 28](https://docs.google.com/presentation/d/1V3v3-g6pAR8oMhk0Az-pheFM7PRCxJO9OsonH7STi5I/edit#slide=id.g20885a66385_1_215), and tried to distill principles by which they were selected on [slide 25](https://docs.google.com/presentation/d/1V3v3-g6pAR8oMhk0Az-pheFM7PRCxJO9OsonH7STi5I/edit#slide=id.g20885a66385_1_215). Since my proposal, Sam has explained why overlays are superior to headers; this means my proposal should be updated to use his better construct. He's also laid out his ESSR thinking. This replaces the confidentiality layer in my proposal, and it also simplifies the privacy mechanisms I wanted. And it explains how to provide the functionality it was imagining for the `src` header that my proposal talked about.

I am not yet satisfied that we have an answer for several of the other "authentic intentions" I talked about. I will have to explain those in a different post, and/or verbally on one of our calls.

**Reply by rodolfomiranda** [reply]
At the proposed solution on [slide 28](https://docs.google.com/presentation/d/1V3v3-g6pAR8oMhk0Az-pheFM7PRCxJO9OsonH7STi5I/edit#slide=id.g20885a66385_1_215), you include three fields marked as `required` (source, signature and schema), and the rest are `optional`.  Should the optional ones be defined in this proposed TTMP layer to satisfy the "other authentic intentions" or can they be left to the other layer since they are marked as optional?

The fields marked as required are covered in all proposal except the `schema` that probably is the easiest to resolve. 
Anyway, further explanation on the missing gaps will help a lot.

**Reply by dhh1128** [reply]
There is a difference in my mind between *requiring* a field (which I DON'T advocate) and *defining* a field (which I DO advocate). Basically, I want the base layer to say, "Here is how to model &lt;semantic concept X&gt;. If you don't need it, leave it out -- but don't reinvent it. And further, if you omit it, here is exactly the semantic that MUST be attributed to a message that omits it." Essentially, this means that TSP eliminates ambiguity about that particular topic.

Moving the *definition* of these fields to a different and optional layer means that the base is ambiguous with respect to whatever issue the field cares about. It is the ambiguity that bothers me, because I think it invites false assumptions, and because it makes the combination of different trust tasks problematic by preventing the free flow of semantics between them. One task might need a semantic that another doesn't support.

If we made the separate layer required (meaning the definition of a semantic is required and is done in that separate layer, but NOT that the field becomes required), the ambiguity would be gone, and I could be happy. But if the layer is required, and all we're really doing is adding a definition (we're NOT adding a dependency on impls by *defining* something; we only add that dependency when we actually *use* it), this begs the question of why it's a separate layer (in my mind, anyway).

**Reply by SmithSamuelM** [reply]
@dhh1128 
>What I want is a future where ToIP is loud and proud that A) those missing things are also essential to any sophisticated trust strategy; B) there's a canonical way to achieve them. If we accomplish A and B, and those things are in a different layer from a some other layer, then I can be happy. But to the extent that we create the extra layer only to make the extra layer an optional thing, and a peer to current or future constructs from arbitrary inventors that want to think about trust differently, IMO we are not building something that deserves the name "trust spanning protocol." We are instead providing a "few key ingredients of trust" protocol.

I in general agree with your sentiment expressed above.  My preference is that we have another overly above the PAC overlays that provides canonical ways of conveying what I label as "transitive trust". I think that "authentic intentions" can be broken down into properties that can be conveyed with trust. I.e. the properties themselves are preserved across trust tasks. I.e. the transitivity crosses trust task boundaries. That means that the end of a sequence of tasks can verify back to the other end. Another way of saying this is that we want end-to-end transitivity of trust.

 I think where there is some incoherence is that composing tasks that do not preserve end-verifiablity  is not transitive trust. It breaks end-to-end transitivity, Instead, It is can be viewed as a form of  "hop-wise" trust in that previous hops are opaque to the next hop. If all we need is "hop-wise" trust, then any composition engine is just as good as any other and we are just canonicalizing a taks composition engine that seems arbitrary to me.  If we didn't add transitive trust to the composition because we broke end-verifiability then we failed.  With hop-wise trust the middle matters, and we have to trust the middle. Which is less trustable because all the middle is exposed to attack. And any successful attack on any of the middle tasks breaks trust between the ends. This is antithetical to the security principle of end-verifiability which in turn is essential to zero-trust computing architectures.

Where this gets sticky is that different programming paradigms have different concepts of proper encapsulation. For many of the most common programming paradigms, this type of transitivity breaks those encapsulation principles.  But there are well-known but less popular programming paradigms, namely FBP (flow based programming), CP (component programming), and DFP (Data flow programming) used heavily in robotics and intelligent control systems, where such end-to-end transitivity is not antithetical to the paradigm. 

 So from my perspective we first need to better define how "trust" properties are transitive in that they preserve end-verifiability (and are appropriately modulated by confidentiality and privacy) across a set of trust tasks.

**Reply by dhh1128** [reply]
>I think where there is some incoherence is that composing tasks that do not preserve end-verifiablity is not transitive trust

Yes, this is one of the sticking points. Not because I disagree with your conclusion; I am well aware of at least some of the limits of transitivity that you're probably concerned about. Rather, I think it's because some things are desirable for trust even without transitive end-verifiability.

**Reply by rodolfomiranda** [reply]
@dhh1128 , follow up question. Based on the ESSR approach, where do you think those additional defined but optional properties should reside? inside the cipher text or signed as plaintext?

**Reply by dhh1128** [reply]
It should be possible for the final recipient(s) to prove that the properties were added by the sender -- and if confidentiality is being added by encryption, it should be impossible for anyone other than the final recipient(s) to know that the metadata has been added or what values the metadata carries.

**Comment by darrellodonnell** [comment]
> Why Separate the TTMP from the TSP?

I'll throw in a simple commentary. We have been running on the basis that the IP layer was the narrow waist that led to success of IP networking. I disagree.

Before TCP/IP really took off things were ugly. Very ugly. Once TCP/IP started to gel we saw two camps:
1. Those that felt TCP/IP was sufficient.
2. Those that didn't.

Camp 1 drove the real adoption while Camp 2 bifurcated further
2.1 Those that slid in UDP
2.2 Those that continued to do their own thing and have largely vanished or slid back under UDP.

TCP/IP was the success factor. Allowing UDP for those special cases where TCP/IP was either too heavy or didn't quite fit helped. 

A naked address, while required, was not sufficient. 

To me we MUST have communication - an equivalent to TCP and UDP. That can be layered on top of the addressing (IP layer) but the community will not benefit from inconsistent approaches to communicating.

**Reply by talltree** [reply]
@darrellodonnell I 100% agree that it took the combination of TCP and IP to solve the problem and win the day, and in fact one of my main motivations in proposing this potential two-level model was to see if it was the **combination** of those two levels that could solve the problem and win the day for us.

What that effectively means, from my POV, is that it doesn't matter whether it is two levels or one, what matters is whether the solution actually solves the problem (works) AND wins the day (is successfully adopted).

**Comment by TelegramSam** [comment]
This is slightly tangential to the main discussion here, but it led me to the details of a practical approach that respects the layers in spirit.

Sam Smith's work has been focused on the encryption and signing aspect of the protocol. Discussions have highlighted the need for more than that to be specified, to make use as a foundational protocol practical. Here are the basics of what I think might be a good approach to discuss during our consolidation phase.

1. All Identifiers are DIDs. A subset of DID methods can be chosen (AIDs for example)
2. Use a close-as-possible adaption of DIDComm's transport and routing structure, accommodating item (3) in this list.
3. Use Sam Smith's Encryption and Signing stuff
4. Message Serialization uses CESR
5. Message and Protocol Semantics use DIDComm-defined semantics.

Even though we are not in agreement that DIDComm did everything right, this approach builds, from the bottom-up a better structure. It constrains our work to figuring out the details of the encryption and signing stuff. It allows reaching a spec faster that is actually usable (version 0.1?). After we get to this point, we can continue to discuss the replacement  of the higher level (DIDComm based) message semantics and replace them in a later version.

By taking such an approach, we constrain our work and discussions in the short term to encryption and signing, without preventing the improvement of higher layers as we progress to that point.

**Reply by rodolfomiranda** [reply]
that approach can also lead to split into subgroups (parallelism). For example one group can focus on creating a new version of DIDComm with CESR serialization and creating a new CESR TSP code table while other group focus on defining the required semantic properties.

**Reply by talltree** [reply]
@TelegramSam As I just [posted above](https://github.com/trustoverip/trust-spanning-protocol/discussions/43#discussioncomment-5598114), the point of this proposal was not so much to advocate for two layers as it was to advocate for a way to achieve alignment among the initial proposals such that we could move into the Working Draft stage. If it can be done by proceeding as you suggest with a single protocol, I'm all for it.

I'm not enough of an expert in DIDComm V2 to fully understand your five steps (except for the first one, which I've already endorsed [here](https://github.com/trustoverip/trust-spanning-protocol/discussions/36#discussioncomment-5588040) in the separate discussion you started on that topic). But since you are such an expert (as co-chair of the DIF DIDComm WG), might I interest you in starting another discussion thread to flesh out in some more detail what you believe those other four steps would look like?

**Comment by jospencer-460** [comment]
I support the logical segregation of these responsibilities. Whether they are "visible" to the consumers of the (broader) TSP or not is a question.

The important point is that the connection establishment protocol must define on what "trust basis" the connection will operate. This includes DID methods, governance framework(s), technologies, data encoding standards, supporting services (VDRs/TRs - which should be inherited from the governance frameworks, but might not be), intermediaries,  OCAs , etc.

The demand on the connection requirement will be made by the initiating (or inheriting) trust task, or decided upon by the TTMP in conjunction with the TSP, where this can be interpreted accordingly. All access to the messaging component (TSP) will be through the TTMP (or whatever we call it).

**Reply by talltree** [reply]
@jospencer-460 I like your phrase, "logical segregation of these responsibilities". I completely agree that, depending on the design, it may not make much difference if they are covered in two layers or one.

But whether that "logical separation" is actually helpful in the design — and if so where we should "draw the line" as to what is required "below the line" and what can be moved "above the line" — is a key topic to discuss in today's TSP Workshop.
